libp2p_bridge/grpc_server.py

```diff
@@ -0,0 +1,84 @@
+from concurrent import futures
+import time
+
+import grpc
+
+from libp2p_bridge.config import (
+    RPC_SERVER_LISTEN_IP,
+    RPC_SERVER_PORT,
+)
+from libp2p_bridge.message import (
+    Collation,
+    CollationRequest,
+    MsgType,
+)
+
+import libp2p_bridge.github.com.ethresearch.sharding_p2p_poc.pb.event.event_pb2 as event_pb2
+import libp2p_bridge.github.com.ethresearch.sharding_p2p_poc.pb.event.event_pb2_grpc as event_pb2_grpc
+
+
+def make_response(status):
+    response = event_pb2.Response()
+    if status:
+        response.status = event_pb2.Response.SUCCESS
+    else:
+        response.status = event_pb2.Response.FAILURE
+    return response
+
+
+def handle_new_collation(collation):
+    return bytes(True)
+
+
+def handle_collation_request(collation_request):
+    c = Collation(collation_request.shard_id, collation_request.period, b"fake")
+    return c.to_bytes()
+
+
+type_msg_map = {
+    MsgType.Collation: (Collation, handle_new_collation),
+    MsgType.CollationRequest: (CollationRequest, handle_collation_request),
+}
+
+
+def dispatch(msg_type, data_bytes):
+    if msg_type not in type_msg_map:
+        return b""
+    msg_cls, handler = type_msg_map[msg_type]
+    deserialized_msg = msg_cls.from_bytes(data_bytes)
+    return handler(deserialized_msg)
+
+
+class GRPCServer(event_pb2_grpc.EventServicer):
+
+    def Receive(self, request, context):
+        response = make_response(True)  # Request succeeded
+        ret_bytes = dispatch(request.msgType, request.data)
+        receive_response = event_pb2.ReceiveResponse(
+            response=response,
+            data=ret_bytes,
+        )
+        print("Receive: request={}, response={}".format(request, receive_response))
+        return receive_response
+
+
+def run_grpc_server():
+    # TODO: should confirm how many workers to use
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
```

Comment:
Am I reading this right that every "in-flight" request (a request sent, with no response received) will hang one thread? So this would limit to 10 in-flight requests.  I'm not really sure what's reasonable, but that sounds a little low. It's worth thinking about what happens if a few of your clients go offline or just time out (hanging a thread for a while).

Reply:
> Am I reading this right that every "in-flight" request (a request sent, with no response received) will hang one thread? So this would limit to 10 in-flight requests. I'm not really sure what's reasonable, but that sounds a little low.

Yes, it is. I set this number just as a placeholder since I wasn't sure of how many threads to use. Should I just set it as None(or don't set this parameter), let it be `5 * num_of_processor`? (ref: [doc](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor)). This number seems more reasonable to me. However, might still be a relatively small number.